### ISSUE TITLE
Make the BareFunctionSegment dialect configurable

### ISSUE THREAD

**Issue by alanmcruickshank** [sqlfluff/sqlfluff#430]
This is an additional improvement on top of #424 . If someone wants to pick this up, please read the backstory of this feature there.

The `BareFunctionSegment` is currently defined in the ANSI dialect thus:

```python
BareFunctionSegment=ReSegment.make(
    r"current_timestamp|current_time|current_date", name="bare_function", type="bare_function"
),
```

This works perfectly if all bare functions are the same in every dialect, but doesn't support dialect configuration very well. 

Similar functionality has already been implemented with for the `DatetimeUnitSegment` using the `sets` and `SegmentGenerator` functionality of the dialect as follows:

```
ansi_dialect.sets('datetime_units').update([
    'DAY', 'DAYOFYEAR', 'HOUR', 'MILLISECOND', 'MINUTE', 'MONTH',
    'QUARTER', 'SECOND', 'WEEK', 'WEEKDAY', 'YEAR'
])

...

DatetimeUnitSegment=SegmentGenerator(
    lambda dialect: ReSegment.make(
       r"^(" + r"|".join(dialect.sets('datetime_units')) + r")$",
        name='date_part', type='date_part')
)
```

The goal of this issue is to update the implementation of `BareFunctionSegment` to be implemented in the same way. Bonus points are available for implementing a couple of dialect specific bare functions at the same time with tests.

**Comment by barrywhart** [comment]
Is this potentially related to #357? The keywords being discussed there are not exactly bare functions, but they do function similarly in some ways.

**Comment by barrywhart** [comment]
Closed via #665
